Desktop_Interface/isobufferbuffer.cpp

```diff
@@ -1,49 +1,114 @@
 #include "isobufferbuffer.h"
 
-isoBufferBuffer::isoBufferBuffer(uint32_t length)
-    : bufferLength(length)
-{
-    mid = bufferLength/2;
-    buffer = (char *) malloc((bufferLength * 3) / 2);
+/* isoBufferBuffer is implemented as two consecutive, duplicate,
+ * ring buffers. the effect of this is that we are able to hand
+ * out pointers to pieces of contiguous memory representing the
+ * last N inserted elements (for some N <= capacity()) even
+ * when we looped back to the begining of the ring buffer less
+ * than N insertions ago
+ *
+ *     There are some differences with the original implementation
+ * functionality-wise.
+ *     Where the original implementation allowed queries half as
+ * long as the length passed in, and allocated a buffer 1.5 times
+ * the length passed in, this version allows queries of length up
+ * to the length passed into the constructor and allocates a
+ * buffer 2 times the length passed in.
+ *     Overall, this means that in contrast to the original
+ * implementation which only allowed queries up to a third as
+ * long as the allocated buffer, we can now do queries as long as
+ * half of the allocated buffer, which is a notable improvement.
+ */
+
+
+// TODO: go through the usages of this class and:
+// 1. adapt code to use the new interface and remove the old one
+// 2. adjust the size of the requested buffer to accomodate
+// the improved memory efficiency of the new algorithm
+
+isoBufferBuffer::isoBufferBuffer (
+	uint32_t length
+) :
+	data_(std::make_unique<char[]>(length*2)),
+	capacity_(length)
+{}
+
+// Adds a character to the end of the buffer
+void isoBufferBuffer::insert ( char c ) {
+	// Add character to first half of the buffer
+	data_[top_] = c;
+	// Then to the second
+	data_[top_+capacity_] = c;
+
+	// Loop the buffer index if necessary and update size accordingly
+	top_ = (top_ + 1) % capacity_;
+	size_ = std::min(size_ + 1, capacity_);
+}
+
+void isoBufferBuffer::insert ( char const * s ) {
+	while ( *s )
```

Comment:
It's generally better to be explicit about comparisons, rather than relying on the fact that specific constructs (e.g. nullptr, null terminator on string) is zero.
It won't affect behaviour on any remotely decent compiler, but it makes your intentions clearer to other people reading the code.
In this case, a better way of writing it would be "while(*s != '\0')".